### PR TITLE
fix: assign commits to closest tag

### PR DESCRIPTION
**Pull request by DaniPopes**
## Description

Previously, when creating releases (assigning commits to tags), a new release was considered only when a tagged commit is encountered. This works fine in most cases, however, when used together with `--include-path`/`--exclude-path`, which may filter out one of these tagged commits, the entire tag would be skipped too.

This PR makes sure tags are never skipped by looking up the closest tag instead of trying to find an exact match for a given commit when creating new releases.

Additionally, this PR also fixes the issue where the first (root) commit of the repository was never included in the output when `--include-path`/`--include-path` was used.

Both of these issues are described in https://github.com/orhun/git-cliff/issues/208.

Based on https://github.com/orhun/git-cliff/pull/709 to minimize conflicts.

## Motivation and Context

Fixes https://github.com/orhun/git-cliff/issues/208.

## How Has This Been Tested?

`cargo test` + manual testing on repositories I use git-cliff on with `--include-path`, and with the script provided in the issue.

## Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other 

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly.
- [x] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [x] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by DaniPopes** [comment]
Keeping as draft as existing tests pass, but I still have to add new ones from the linked issue and some more edge cases I have in mind

File: git-cliff/src/lib.rs [hunk 1]
```diff
@@ -324,6 +306,42 @@ fn process_repository<'a>(
 	Ok(releases)
 }
 
+fn fill_release(
```
Comment (orhun) [inline git-cliff/src/lib.rs:309]: Do we need separate functions for these? I feel like they are additional complexity.

I'm okay with it if we test them though.

**Comment by orhun** [comment]
Hey @DaniPopes are you still interested in working on this?

**Comment by DaniPopes** [comment]
Hey, sorry I haven't been able to allocate more time on this, and I'll be off for a few weeks.
Might come back to this in a month or so

**Comment by orhun** [comment]
ayo what's the progress on this 👀

**Comment by orhun** [comment]
closing due to inactivity

**Comment by zerosnacks** [comment]
Hi @DaniPopes considering we currently point to this PR for our crate release workflow I think it make sense to attempt to get this re-opened + upstreamed. The PR looks quite large but I think it should be ready to be included right?

**Comment by orhun** [comment]
Hey @zerosnacks, I unfortunately couldn't get a response from @DaniPopes about this...

Would you be possibly interested in finishing this up in another PR?

**Comment by zerosnacks** [comment]
Hey @orhun, I work together with @DaniPopes so I'll get in touch directly and see if we can get this over the line :smile: ; figured I would ping them here for context
